### PR TITLE
AML clock timing adjustments

### PR DESCRIPTION
**Pull request by peak3d**
Timing adjustments to get AML clock better synced to kodi DVDClock

## Description
This PR does mainly 2 things:
- wait for AML decoder start until enough packets are feeded
- adjust AML clock to be never behind kodi clock (can lead to stutter) and not to much behind kodi clock(can lead to stutter as well)

## Motivation and Context
MPEG2 videos delivered from TVHeadend do not play without stutter.

## How Has This Been Tested?
Runtime tests 23.97fps, 25 fps , 50fps live-tv and streams on local drive.

## Types of change
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp [hunk 1]
```diff
@@ -1859,24 +1868,37 @@ int CAMLCodec::Decode(uint8_t *pData, size_t iSize, double dts, double pts)
...
-    rtn, m_cur_pictcnt, (float)m_cur_pts/PTS_FREQ, (float)am_private->am_pkt.lastpts/PTS_FREQ, GetTimeSize(), GetDataSize());
-*/
+
+  if (g_advancedSettings.CanLogComponent(LOGVIDEO))
+    CLog::Log(LOGDEBUG, "CAMLCodec::Decode: ret: %d dts_in: %0.6f, pts_in: %0.6f, ptsOut:%0.6f"
+      ,rtn
+      ,static_cast<float>(dts)/DVD_TIME_BASE
+      ,static_cast<float>(pts)/DVD_TIME_BASE
+      ,static_cast<float>(m_cur_pts)/PTS_FREQ
+    );
```
Comment (AlwinEsch) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp]: Is it not a bit better in design to have the commas on end? ;)

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp [hunk 2]
```diff
@@ -2005,10 +2027,16 @@ void CAMLCodec::Process()
...
+    {
+      usleep(1000);
+      continue;
+    }
+
     if (m_dll->codec_poll_cntl(&am_private->vcodec) < 0)
     {
       CLog::Log(LOGDEBUG, "CAMLCodec::Process: codec_poll_cntl failed");
-      Sleep(10);
+      usleep(10000);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp]: why?

Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp]: Ah, capital S, ok, mixed it up and thought it is seconds and therefore an bug.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp (hunk 1 above)
Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp]: form ergonomic pov its easier at the beginning, because you can comment out things better, especially the last line. But Ok for me to change if you like it more

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.cpp [hunk 3]
```diff
@@ -21,43 +21,91 @@
 #include "RendererAML.h"
```
Comment (peak3d) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.cpp]: I removed the inheritance to LinuxRenderGLES because after some mails with AMLogic it will never be a solution and will never be targetted to use GPU for rendering. GPU will never be working with 4K and DRM.

**Comment by stefansaraev** [comment]
v18 please. this is pretty intrusive change and breaks (at least) my uhd astra19 samples on s905.

fiy @codesnake @kszaq @wrxtasy

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp [hunk 4]
```diff
@@ -1895,7 +1895,7 @@ int CAMLCodec::Decode(uint8_t *pData, size_t iSize, double dts, double pts)
     m_last_pts = m_cur_pts;
     m_cur_pts = decode_pts;
   }
-  if ((rtn & VC_PICTURE)==0 || timesize < 1.0)
+  if (((rtn & VC_PICTURE)==0 && timesize < 2.0) || timesize < 1.0)
```
Comment (hudokkow) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp]: Space around operators, please.

**Comment by mcaptur** [comment]
latest update does not compile

ake[2]: Leaving directory '/home/mark/src/LibreELEC.tv/build.LibreELEC-WeTek_Play.arm-8.0-devel/kodi-17.0-beta6-a37127e/xbmc/guilib'
In file included from AMLCodec.cpp:23:0:
AMLCodec.h: In constructor 'CAMLCodec::CAMLCodec()':
AMLCodec.h:101:16: warning: 'CAMLCodec::m_prefill_state' will be initialized after [-Wreorder]
   PREFILLSTATE     m_prefill_state;
                ^~~~~~~~~~~~~~~
AMLCodec.h:75:8: warning:   'bool CAMLCodec::m_opened' [-Wreorder]
   bool             m_opened;
        ^~~~~~~~
AMLCodec.cpp:1410:1: warning:   when initialized here [-Wreorder]
 CAMLCodec::CAMLCodec()
 ^~~~~~~~~
In file included from AMLCodec.cpp:23:0:
AMLCodec.h:84:11: warning: 'CAMLCodec::m_last_pts' will be initialized after [-Wreorder]
   int64_t          m_last_pts;
           ^~~~~~~~~~
AMLCodec.h:80:20: warning:   'volatile int64_t CAMLCodec::m_cur_pts' [-Wreorder]
   volatile int64_t m_cur_pts;
                    ^~~~~~~~~
AMLCodec.cpp:1410:1: warning:   when initialized here [-Wreorder]
 CAMLCodec::CAMLCodec()
 ^~~~~~~~~
AMLCodec.cpp: In member function 'int CAMLCodec::Decode(uint8_t*, size_t, double, double)':
AMLCodec.cpp:1906:33: error: 'RD_ONLY' was not declared in this scope
     int fd(open("/dev/amvideo", RD_ONLY);
                                 ^~~~~~~
make[2]: *** [/home/mark/src/LibreELEC.tv/build.LibreELEC-WeTek_Play.arm-8.0-devel/kodi-17.0-beta6-a37127e/Makefile.include:97: AMLCodec.o] Error 1
make[2]: Leaving directory '/home/mark/src/LibreELEC.tv/build.LibreELEC-WeTek_Play.arm-8.0-devel/kodi-17.0-beta6-a37127e/xbmc/cores/VideoPlayer/DVDCodecs/Video'
make[1]: *** [Makefile:432: xbmc/cores/VideoPlayer/DVDCodecs/Video/Video.a] Error 2
make[1]: *** Waiting for unfinished jobs....

**Comment by peak3d** [comment]
@mcaptur @hudokkow both issues fixed

**Comment by mcaptur** [comment]
works fine here :)

**Comment by mcaptur** [comment]
CPP     xbmc/interfaces/python//home/mark/src/LibreELEC.tv/build.LibreELEC-WeTek_Play.arm-8.0-devel/kodi-17.0-beta6-ee22e6e/xbmc/interfaces/python/generated/AddonModuleXbmcplugin.o
AMLCodec.cpp: In function 'int check_in_pts(am_private_t*, am_packet_t*)':
AMLCodec.cpp:636:3: error: expected ')' before '{' token
   {
   ^
AMLCodec.cpp:644:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
 ^

**Comment by peak3d** [comment]
@mcaptur you have been too fast, I fixed this one already some minutes after initial commit

**Comment by mcaptur** [comment]
yep compiles now :)

**Comment by Mister-XY** [comment]
When will this Pull request add to the branch?

**Comment by codesnake** [comment]
@Mister-XY Unfortunately this PR still has problems playing some types of videos, so I think it is too early to merge it.

**Comment by peak3d** [comment]
@codesnake can you pls provide an example?

**Comment by codesnake** [comment]
@peak3d http://sprunge.us/IiId

**Comment by codesnake** [comment]
@peak3d Any comment on the log I posted?

**Comment by peak3d** [comment]
@codesnake interesting log.

00:46:22.131 T:2857395104   DEBUG: CAMLCodec::Decode: ret: 2 dts_in: 0.937488, pts_in: 1.017488, ptsOut:0.000000, amlpts:0 vfs:[65-1--306045168--297715856] timesize:0.02
00:46:22.132 T:2857395104   DEBUG: CVideoPlayerVideo::CalcDropRequirement - hurry: 0
00:46:22.135 T:2857395104   DEBUG: CAMLCodec::Decode: ret: 0 dts_in: 0.957488, pts_in: 1.037488, ptsOut:0.000000, amlpts:0 vfs:[65-1--306045168--297715856] timesize:7.00

the timesize value changes from one to the other frame from 0.2 to 7.0.
I'll add some more debug in GetTimeSize() to get an idea what's happening there.

Is this issue reproducable for you? ( I was unable so far to get an similar issue)

**Comment by peak3d** [comment]
@codesnake there was indeed one initialization missing / wrong.
This is fixed and additionally some debug output if timesize leaves expected limits.

Cold you pls. check again and if the issue still appears provide the log again? Thx!

**Comment by codesnake** [comment]
@peak3d Unfortunately nothing changed. Still cannot open any HD channel: http://sprunge.us/jKKd

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp [hunk 5]
```diff
@@ -1391,6 +1391,8 @@ CAMLCodec::CAMLCodec()
   am_private->vcodec.cntl_handle        = -1;
   am_private->vcodec.sub_handle         = -1;
   am_private->vcodec.audio_utils_handle = -1;
+  am_private->vcodec.has_audio = 0;
+  am_private->vcodec.has_sub = 0;
```
Comment (codesnake) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp]: Anyway these already set to zero by memset above.

**Comment by peak3d** [comment]
@codesnake I have removed the codec_get_video_cur_delay_ms calculation as it returns unwanted things.
The timesize calulation should be working now for the material you have.

**Comment by koenkooi** [comment]
Does this improve the 2x and 4x fast forward behaviour?  With kszaqs jarvis builds on s905 I get one second of proper fastforward and then the picture slows down and starts skipping ahead.

**Comment by kszaq** [comment]
@koenkooi Jarvis is much different than Krypton regarding to Amlogic a/v sync.

**Comment by hephooey** [comment]
After updated this pr today I cannot play any twitch live/video with kodi anymore, you can see part of the output [here](http://pastebin.com/89ZB1D2u). I remember twitch somewhat worked (video still stuck from time to time) at 1b37a4f.

**Comment by peak3d** [comment]
@koenkooi 
> Does this improve the 2x and 4x fast forward behaviour?

Unfortunately not, its still the AML clock we use, but development has not stopped and we try together with AMLogic to come to an better concept for kodi.

@hephooey 
>After updated this pr today I cannot play any twitch live/video with kodi anymore

Would be gread if you can provide something wich lets me reproduce it.
From log 2 seconds Input buffer desn't lead to an decoded frame

**Comment by hephooey** [comment]
@peak3d, the one I used to generate the output is https://vod135-ttvnw.akamaized.net/v1/AUTH_system/vods_b2fe/lordzephyrhs_23848275840_559188128/chunked/index-dvr.m3u8. Though twitch change the url from time to time so it might not be valid when you test it. There is also a twitch plugin for kodi in the official repo, you can use it as well.

**Comment by peak3d** [comment]
thanx @hephooey , found the addon and can reproduce...

**Comment by peak3d** [comment]
@hephooey twitch has only a few IDR frames in it's live video streams.
The issue is fixed with the last commit by skipping all NON IDR frames on start.

**Comment by FernetMenta** [comment]
h.264 streams are not required to have IDR frames. you might wait forever with this change.

**Comment by hephooey** [comment]
@peak3d, yes, twitch streams are fine now. But I just found some youtube video stop playing after this new commit, for example: https://www.youtube.com/watch?v=768sISC9ERA

**Comment by codesnake** [comment]
@peak3d Still cannot open any of my HD DVB-S2 channels: http://sprunge.us/JgAi

**Comment by fritsch** [comment]
Tvheadend is not a benchmark that very moment. Use recordings from disk.

It goes that way with hw decoders: get everything running and afterwards
try to cope with tvh.

Am 08.12.2016 2:20 vorm. schrieb "Alex Deryskyba" <notifications@github.com

**Comment by peak3d** [comment]
@codesnake @hephooey : @FernetMenta already commented the issue directly after the push.
To evaluate recovery_count in SEI NAL Units is nothing I do every day so it took some time to get it.

The commit is in now, At least @codesnake's stream should be able to run now. 
(@codesnake: what means the "still not running"? It ran before only seeking was not Ok, wasen't it?)

**Comment by peak3d** [comment]
@fritsch @FernetMenta The implementation assures, that there is no difference due to the changes for existing streams. The search for IDR / SEI has to be enabled by ResetKeyframe(). Without resetting, BitstreamConverter does not run into the new implementation. This is done to avoid extra overhead for codecs wich don't need this extra step.

**Comment by codesnake** [comment]
@peak3d I never told about seeking and there's no seeking in Live TV. When I said I cannot open any HD channel I meant I see only infinite buffering and channel never start to play.

**Comment by peak3d** [comment]
@codesnake fine, then I mixed up things.
Can you pls try the current version? It should work now.

Thank you!

**Comment by codesnake** [comment]
@peak3d Unfortunately still no go. As @fritsch suggested I tried a few recorded samples on WeTek Play and some of them cannot be played. For example this: https://mega.nz/#!JRJFzAiI!_SyzYuspoLVJyXnZHIiaROgw6iw-gu_Dp2ODOMRlgtQ

**Comment by peak3d** [comment]
@codesnake -> fixed. Really great thanx for your patience.... can you try again?

**Comment by hephooey** [comment]
@peak3d, the youtube video I posted yesterday still does not play. You can see the log here: http://pastebin.com/Q2z4pPeH

**Comment by codesnake** [comment]
Live TV now works OK, thanks. But 23.976 content is still stuttery in comparison with Jarvis. To be honest in general I don't see any difference in playback with or without this PR, given the fact it adds 80K of changes.

**Comment by peak3d** [comment]
Its more a cleanup than an improvement (beside the fact that we have many streams now running wich doesn't run before). For me (and therefore many others wich use kodi 17 on odroid's) this PR makes kodi17 working, it wasn't before (ask my wife). No idea whats the difference on your system and to be honest, I don't care about the fact that it runs for you. I tend to look for issues why anything des NOT run.

To get the last piece working, could you pls. so nice and post a debug log of 23.976 on 24Hz Monitor?

**Comment by hephooey** [comment]
@codesnake, one positive example of this pr comparing with Jarvis on an Odroid C2 is the video posted [here](http://forum.kodi.tv/showthread.php?tid=258707&pid=2450781#pid2450781). On Jarvis it is a pain to watch the jittering pictures, while with this PR everything is smooth. I do not know if the situation are similar on other amlogic based devices.

**Comment by codesnake** [comment]
@peak3d Besides your wife there are a lot of other users, and besides Odroid C1 there are a lot of other Amlogic-based devices. I think you PR improve some things, but you include a lot of changes here that at least shouldn't be a part of this PR.

Can you please provide a sample that doesn't work without your PR?

Regarding 23.976 you'll not see anything in Kodi log because it skips frames kernel-side.

**Comment by peak3d** [comment]
@hephooey I downloaded the YT video using different tools (YT addon tells be 403 forbidden) but all downloaded streams worked fine ( I downloaded 720p as its the one you tried). If possible pls try to paste the URL wich is in the kodi.log file in the browser or curl or whatever to get the stream you really see. I need some seconds / megabyte of the stream for testing.

@codesnake Deluxe Music over TVH (but the issue is for all mpeg2 Live-Tv). But, pls. don't let us discuss. You know as I know that the OMX time-window is too big to allow AV sync. Do something wich shortly interupts kodi and you're unsync.

Sentenses as "you did changes wich should not be done" are to general for me (and are ignored). The PR is out for several weeks and it has been enough time to comment these places in git. If you know that there are things wrong. pls comment and let us know why. I'm always open for improvements.

**Comment by hephooey** [comment]
@peak3d, I use the youtube plugin and the firefox extension "send to xbmc" to play that video in kodi. I hope you can reproduce the problem this way. Meanwhile I never tried to download the video and play it before. I will try to download it with youtube-dl and test it locally when I got home.

**Comment by hephooey** [comment]
@peak3d, Just got home and tried again, it seems everything works fine now. Must be some problem with my internet this morning..., sorry about the false alarm.

**Comment by codesnake** [comment]
@peak3d Is there any particular reason why you derived CRendererAML from CBaseRenderer? From my tests it seems doesn't add any value, just more code. I tried some of the previous variants of your PR and I got best results of 23.976 playback in 24hz mode when CRendererAML was in this state: http://pastebin.com/JaxXStyN

**Comment by peak3d** [comment]
@codesnake the current implementation uses only the yuv buffers from GLES Renderer.
This includes the count of render buffers as well. To have the possibility to play around with the # of buffers its much easier to have it in a class where it does not affect the GLES implemenation.

After some discussions with amlogic we figured out, that it will never ever be the way to display the images with GPU, reason is that they don't focus on that and current issues (DRM / 4K) will most likely never ever be possible over GPU.

Long story short, its a cleanup step and should not affect the workflow at all. We dont use anything from GLES in the AML render pipeline and therefor it should not be inherited from GLES Renderer.

I'll look into the log now, I'm sure that we'll be able to get 23.976 / 24 running soon.

**Comment by peak3d** [comment]
@hephooey yes, I tried as well using firefox plugin and it runs well.

**Comment by mcaptur** [comment]
fails to compile with latest updates

patching file xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.h
patching file xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.cpp
patching file xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp
Hunk #1 succeeded at 1609 (offset 56 lines).
Hunk #2 succeeded at 1744 (offset 56 lines).
Hunk #3 succeeded at 1843 (offset 56 lines).
patching file xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.cpp
patching file xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp
Hunk #1 succeeded at 1767 (offset 56 lines).
Hunk #2 succeeded at 1877 (offset 56 lines).
patching file xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.h
Makefile:9: recipe for target 'release' failed
make: *** [release] Error 1

**Comment by peak3d** [comment]
@mcaptur the PR is rebased on current master (K18).

**Comment by frznGT** [comment]
@peak3d Thanks for your work on this, many videos run more smoothly for me now.  However, I have run into an instance where a VC-1 coded video stalled after about 16 seconds. I checked and found that the vbuf was nearly empty while the timesize remained above 2. Here is the log section: [pastebin.com/4i4dAFbF](http://pastebin.com/4i4dAFbF)

**Comment by hephooey** [comment]
@peak3d, The PR can play all of the local video I tested now, but I still have the video freeze pretty consistently when I watch twitch streams. I think those streams usually are imperfect, contain jumping pts or something like that. It happens very often when the streamer has a bad internet connection. But it happens to almost all the stream if I wait a couple hours. As an example you can test [this stream](https://www.twitch.tv/teamsp00ky/v/107514581) (just type in the url in the twitch addon). 
The stream freeze at many places, for example at 51:42.
kodi.log just flood with messages like:
`CAMLCodec::GetTimeSize limits exceed: avdts: 279534870  cur_pts: 279084420`
and the stream never recover.
Jarvis played this stream without problem, though I did notice in dmesg I have many
`codec:V-discontinue,pcr=326295679,vpts=326289544,apts=0,diff_pts=326648374,jump_Pts=358830`
